### PR TITLE
fix: suppress console windows for git subprocess calls on Windows

### PR DESCRIPTION
**Pull request by ya1yi2fa3**
## Problem

On Windows, short-lived `git.exe` children launched by the WebUI backend can create visible console windows (or Windows Terminal tabs). Update checks and periodic revision/status probes make this especially disruptive.

## Root cause

The backend had inconsistent Windows subprocess handling. `api/workspace_git.py` had a local `CREATE_NO_WINDOW` helper, while Git subprocesses in `updates.py`, `rollback.py`, `workspace.py`, `worktrees.py`, and `agent_runtime.py` either duplicated that helper or omitted the flag entirely.

The frontend's generic `api()` timeout is 30 seconds. `/api/updates/check` can perform bounded network fetches for both the WebUI and Agent repositories, so its callers need a larger single-attempt latency budget. `api()` does **not** retry ordinary timeouts unless `retryTimeouts: true` is explicitly supplied; none of these callers enables that behavior.

### Backend

- Added dependency-light `api/subprocess_utils.py` with one `windows_hide_flags()` implementation.
- Applied `creationflags=windows_hide_flags()` to all 13 direct production Git `subprocess.run` boundaries:
  - `api/agent_runtime.py`: 3
  - `api/rollback.py`: 4
  - `api/updates.py`: 1
  - `api/workspace.py`: 1
  - `api/workspace_git.py`: 2
  - `api/worktrees.py`: 2
- Preserved every existing timeout, captured stdout/stderr setting, environment, and error path.
- Added an AST inventory regression covering all production Python modules, plus real helper behavior tests for a Win32 sentinel and the POSIX zero no-op.

### Frontend

- Kept only the two endpoint-specific `/api/updates/check` latency budgets at 300 seconds (boot background check and manual check).
- Reverted the explicit 30-second settings timeout, unsupported 60-second settings timeouts, and the 60-second `/api/git-info` timeout.
- Updated the three boot/order source-contract tests to accept an optional `api()` options object without weakening their ordering assertions.
- Updated the timeout test oracle to assert `300000` and document the values as single-attempt latency budgets, not retries.

## Verification

- Reviewer-named source-contract files: `51 passed`
- Git subprocess inventory + Agent revision guard: `24 passed`
- Update backend: `76 passed`
- Update checker: `14 passed`
- Worktree backend: `32 passed`
- Workspace Git on this Windows host: `51 passed, 31 skipped, 2 deselected`; the two deselected pre-existing environment cases require a different CRLF Git configuration and Windows symlink privilege. Cross-platform CI remains authoritative for those cases.
- Ruff merge-base added-line gate: `0` new findings
- `npm run lint:runtime`: passed
- Python compile, Node syntax checks, and `git diff --check`: passed
- Windows smoke test: helper value exactly matched `subprocess.CREATE_NO_WINDOW`; a real captured `git --version` returned 0 with stdout intact.

### CODE REVIEW

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
## Summary

The Windows subprocess changes are correct at the eight Git call sites this PR edits, but the fix is incomplete and the current head breaks four existing targeted tests. I would not merge this head yet.

## Blocking findings

1. `api/agent_runtime.py::_read_agent_revision()` still launches three captured Git probes without `CREATE_NO_WINDOW` (`rev-parse --show-toplevel`, `ls-files --error-unmatch`, and `rev-parse --verify HEAD`). That function runs while capturing and rechecking the loaded Agent revision, so affected Windows paths can still open up to three visible Git console windows. This is the same runtime problem the PR claims to fix, outside the four modules currently patched.

2. The sandbox-gated targeted run is red: **165 passed, 4 failed**. The timeout edits invalidate three existing boot/settings source-contract tests and one timeout-contract test:
   - `test_workspace_blank_page_fix.py::...test_boot_sets_profile_default_workspace_in_settings_block`
   - `test_5167_busy_input_mode_boot.py::...test_eager_default_precedes_async_settings_fetch`
   - `test_4961_url_query_prefill.py::test_explicit_session_url_keeps_target_session_with_prefill`
   - `test_api_timeout.py::test_update_flows_keep_explicit_longer_timeouts`

The first three still search the old exact `await api('/api/settings')` shape, and the fourth still requires the prior 60-second manual-update timeout.

There is also a rationale mismatch worth correcting: `api()` retries network `TypeError`s, not ordinary timeout failures unless `retryTimeouts:true` is supplied. None of these calls supplies it, so the larger timeouts are single-attempt latency budgets; they do not prevent a timeout-triggered three-retry cascade.

## Requested changes

- Put the Windows flag logic in one dependency-light helper and use it from `rollback.py`, `updates.py`, `workspace.py`, `worktrees.py`, `workspace_git.py`, and `agent_runtime.py`.
- Add the flag to all three `agent_runtime.py` Git probes while preserving captured stdio and the existing timeouts.
- Update the three ordering tests to accept an API options object while retaining their real ordering assertions, and update the manual-update timeout oracle to the intended value.
- Add an inventory regression that covers every production Git `subprocess.run` boundary, including `agent_runtime.py`, plus Win32-sentinel and POSIX-zero helper cases.
- Describe the frontend timeout changes as latency budgets. Revert the behaviorally inert 30-second explicit setting and any 60-second settings changes that do not have endpoint-specific latency evidence.

I ran the targeted tests only through the mandatory no-network bubblewrap gate after a CLEAN threat scan. No full suite or bare PR-code execution was used.

File: api/updates.py [hunk 1]
```diff
@@ -203,6 +203,17 @@ def _wait_until_restart_safe(poll_seconds: float = 2.0, max_wait_seconds: float
...
 
+def _windows_hide_flags() -> int:
+    """Win32 ``creationflags`` that hide a short-lived console child's window
+    (``CREATE_NO_WINDOW``) without detaching it, so ``capture_output`` still
+    works. Returns ``0`` on non-Windows — the ``subprocess`` default, a genuine
+    no-op. See #5692.
+    """
+    if sys.platform == "win32":
+        return getattr(subprocess, "CREATE_NO_WINDOW", 0)
+    return 0
```
[the remark this answers is not in the document]
Comment (ya1yi2fa3) [reply api/updates.py]: Addressed in 348a36e0: the duplicated platform helpers are replaced by dependency-light `api/subprocess_utils.py::windows_hide_flags()`, and all 13 direct production Git subprocess boundaries now use the shared helper. Added AST inventory coverage plus Win32-sentinel and POSIX-zero behavior tests.

**Comment by ya1yi2fa3** [comment]
Addressed the requested changes in `348a36e0`.

- Centralized the Windows flag in dependency-light `api/subprocess_utils.py` and applied it to all 13 direct production Git `subprocess.run` boundaries, including all three `agent_runtime.py` probes, while preserving stdio capture and timeouts.
- Added an AST inventory regression for every current production Git boundary, plus Win32-sentinel and POSIX-zero helper cases.
- Reverted the inert/default settings and unsupported `/api/git-info` timeout edits; retained only the two endpoint-specific `/api/updates/check` single-attempt latency budgets at `300000`.
- Updated the three ordering tests to accept an optional API options object without weakening their ordering assertions, and corrected the timeout oracle/rationale.

Fresh verification: reviewer-named source contracts `51 passed`; Git inventory + Agent revision guard `24 passed`; update backend `76 passed`; update checker `14 passed`; worktree backend `32 passed`; Ruff added-line gate clean; ESLint runtime guard passed; Python/Node syntax checks and `git diff --check` passed. Independent pre-commit review returned `passed: true` with no security concerns, logic errors, or requirement gaps.

Re-requesting review.

**Comment by ya1yi2fa3** [comment]
@nesquena-hermes The requested fixes are now pushed in `348a36e0`. GitHub rejected the formal reviewer-request API because the fork author lacks `RequestReviewsByLogin` permission on the upstream repository, so I am explicitly requesting re-review here.

**Comment by nesquena-hermes** [comment]
## Summary

Reading exact head `348a36e0017d`, the complete PR diff, the full new helper and inventory test, the full `api/agent_runtime.py`, and both head/master versions of `api/workspace_git.py`, the follow-up addresses the blockers from the previous review. The Windows flag is centralized, all 13 current direct production Git `subprocess.run` boundaries use it, the three Agent-revision probes are included, and the source-contract tests retain their ordering assertions while allowing an API options object. I found no new blocking issue in this head by static inspection.

## Code reference

The shared implementation at `api/subprocess_utils.py:8-18` is dependency-light and keeps non-Windows behavior unchanged:

```python
def windows_hide_flags() -> int:
    """Hide a short-lived console child on Win32 and remain a POSIX no-op.

    ``CREATE_NO_WINDOW`` keeps captured stdout and stderr connected, unlike
    detaching the process. Passing ``0`` elsewhere preserves the subprocess
    default. See #5692.
    """
    if sys.platform == "win32":
        return getattr(subprocess, "CREATE_NO_WINDOW", 0)
    return 0
```

The previously missed boundaries are now covered at `api/agent_runtime.py:45-83`: each of the worktree, tracked-file, and HEAD probes passes `creationflags=windows_hide_flags()` without changing capture, timeout, or error handling. The same helper replaces the local implementation at `api/workspace_git.py:20-24`, and the other ten call sites are visible in `api/rollback.py:102-280`, `api/updates.py:206-229`, `api/workspace.py:1680-1690`, and `api/worktrees.py:17-334`.

## Diagnosis / recommendation

The new AST inventory at `tests/test_git_subprocess_windows_flags.py:17-113` is a useful regression gate: it asserts the exact per-file count, requires the shared helper at every recognized call, and rejects duplicate helper definitions. The Win32 sentinel and POSIX-zero tests exercise the helper itself rather than only checking source text.

The retained frontend changes also match the corrected rationale. `static/boot.js:3454-3458` and `static/panels.js:11936-11957` set a 300-second single-attempt budget only for `/api/updates/check`; they do not enable timeout retries or broaden settings and Git-info deadlines.

One non-blocking maintenance note: `_git_argv_kind()` intentionally recognizes the current direct argv shapes. A future dynamically assembled Git command could evade that classifier, so additions to process-launch code should still update the inventory recognizer rather than only its expected-count map.

## Verification

I performed read-only review only and did not execute contributor code. The diff preserves every existing subprocess timeout, output-capture setting, environment, and return-code path at the edited backend boundaries. On the exact head, the submitted checks are not currently represented by GitHub status checks, so the next maintainer gate should run the focused Windows/helper and source-contract suites before merge.

**Comment by nesquena-hermes** [comment]
## Gate: SAFE TO SHIP at head `348a36e001`

Re-gated fresh at the exact head (rebased clean onto current master):
- **Codex: SAFE TO SHIP, zero regression risk.** All 13 production git subprocess boundaries route through `windows_hide_flags()`; no `Popen`/`check_*`/shell/aliased boundary missed. AST comparison vs `origin/master` confirms argv + every existing kwarg (capture/timeout/cwd/env/check/ordering) is unchanged at every site — `creationflags=` is purely additive. POSIX returns `0` (subprocess default = true no-op on Linux/macOS). Tests are non-vacuous (Win32/POSIX branches + AST call-kwargs + runtime kwargs asserted).
- **Full pytest suite: 14,698 passed / 0 failed.**
- No visible surface (POSIX no-op; the two JS touches are +1 each and pass `node --check`).

This matches the maintainer's prior complete exact-head re-review (no blocker). Queued to ship in the next experimental release. Thanks @ya1yi2fa3.

**Comment by nesquena-hermes** [comment]
Shipped in **exp-v0.52.240** 🎉 — thanks @ya1yi2fa3.

Your `windows_hide_flags()` helper (`CREATE_NO_WINDOW` on Win32, no-op on POSIX) now routes every git subprocess boundary, so short-lived git commands no longer flash a console window on Windows while keeping captured stdout/stderr intact.

Gate at head `348a36e001`: Codex confirmed SAFE TO SHIP with zero regression — all 13 production git subprocess boundaries covered, AST comparison vs master confirmed argv + every existing kwarg unchanged at each site (purely additive `creationflags=`), POSIX no-op verified. Full suite 14,698/0.
